src/core/packages/chrome/navigation/src/hooks/use_responsive_menu.ts

```diff
@@ -32,80 +92,56 @@ interface ResponsiveMenuState {
+export function useResponsiveMenu(isCollapsed: boolean, items: MenuItem[]): ResponsiveMenuState {
+  const primaryMenuRef = useRef<HTMLElement | null>(null);
+  const heightsCacheRef = useRef<number[]>([]);
 
-  const [visibleMenuItems, setVisibleMenuItems] = useState<MenuItem[]>([]);
+  const [visibleMenuItems, setVisibleMenuItems] = useState<MenuItem[]>(items);
   const [overflowMenuItems, setOverflowMenuItems] = useState<MenuItem[]>([]);
 
   const recalculateMenuLayout = useCallback(() => {
-    const menuElement = primaryMenuRef.current;
-    if (!menuElement) {
-      return;
-    }
+    if (!primaryMenuRef.current) return;
 
-    const menuRect = menuElement.getBoundingClientRect();
-    const computedStyle = window.getComputedStyle(menuElement);
-    const itemGap = parseFloat(computedStyle.getPropertyValue('gap')) || 0;
-    const renderedItemHeights = getActualItemHeights(menuElement);
+    // Primary menu
+    const menu = primaryMenuRef.current;
+    const menuHeight = menu.clientHeight;
 
-    const actualItemHeights: number[] = [];
+    // 1. Cache the heights of all children
+    cacheHeights(heightsCacheRef, menu, items);
 
-    for (let i = 0; i < items.primaryItems.length; i++) {
-      if (i < renderedItemHeights.length) {
-        actualItemHeights.push(renderedItemHeights[i]);
-      } else {
-        const avgHeight =
-          renderedItemHeights.length > 0
-            ? renderedItemHeights.reduce((sum, h) => sum + h, 0) / renderedItemHeights.length
-            : isCollapsed
-            ? COLLAPSED_MENU_ITEM_HEIGHT
-            : EXPANDED_MENU_ITEM_HEIGHT;
-        actualItemHeights.push(avgHeight);
-      }
-    }
+    if (heightsCacheRef.current.length === 0) return;
 
-    const calculations: MenuCalculations = {
-      availableHeight: menuRect.height,
-      itemGap,
-      maxVisibleItems: MAX_MENU_ITEMS,
-    };
-
-    const maxVisibleItems = calculateVisibleItemCount(calculations, actualItemHeights);
-
-    // if we need overflow (more total items than can fit), we need to account for "More" button space
-    let adjustedMaxVisible = maxVisibleItems;
-    if (items.primaryItems.length > maxVisibleItems) {
-      const moreButtonHeight = actualItemHeights[0] || EXPANDED_MENU_ITEM_HEIGHT;
-      const availableForRegularItems = calculations.availableHeight - moreButtonHeight;
-
-      const adjustedCalculations = {
-        ...calculations,
-        availableHeight: availableForRegularItems,
-      };
-      adjustedMaxVisible = calculateVisibleItemCount(adjustedCalculations, actualItemHeights);
-    }
+    // Primary menu items
+    const childrenHeights = heightsCacheRef.current;
+    const childrenGap = getStyleProperty(menu, 'gap');
 
-    const { visible, overflow } = partitionMenuItems(items.primaryItems, adjustedMaxVisible);
+    // 2. Calculate the number of visible menu items
+    const visibleCount = countVisibleItems(childrenHeights, childrenGap, menuHeight);
```

Comment:
We discussed a problem with the initial render. The item count is changing, which causes a flicker due to unmounting and remounting. We'll need to follow up to improve